### PR TITLE
Use a shared Breadcrumbs component throughout the app

### PR DESCRIPTION
**Pull request by marekweb**

### CODE REVIEW

**Comment by lguychard** [comment]
TODOs stemming from our last session:
- [ ] Make sure `<Breadcrumbs/>` matches the styling in [figma](https://www.figma.com/file/wxYRk6fiYc7mSGLE1jvz1f/%2311774-Design-templates-for-common-page-types-%5BFinal-review%5D?node-id=0%3A1) 
- [ ] Always add `Home` breadcrumb
- [ ] Use `<Breadcrumbs/>` and `useBreadcrumbs` in place of `RepoHeaderContributionPortal` in `RepoContainer` (may require converting components to function components to be able to use the hook)
    - [ ] routes.tsx: add filePath
    - [ ] RepoRevisionContainer: add revision element
    - [ ] RepositoryBranchesArea: add "branches"
    - [ ] RepositoryCommitsPage: add "commits"
    - [ ] RepositoryComparArea: add "compare"
    - [ ] RepositoryReleasesArea: add "Tags"
    - [ ] RepoSettingsArea: add "Settings"
    - [ ] RepoStatsArea: add "Contributors"
- [ ] Use `<Breadcrumbs/>` for other page types:
    - [ ] Insights
    - [ ] Campaigns
    - [ ] Extensions
    - [ ] Search
    - [ ] Explore
    - [ ] Query builder
    - [ ] Code statistics
    - [ ] Settings
    - [ ] Organization
- [ ] Write & use common header component for other page types

**Comment by lguychard** [comment]
@marekweb I think you might be the one working on this the soonest here. Could you decide what next steps you want to tackle and tell me & @felixfbecker what we should help with?

**Comment by marekweb** [comment]
Starting from the top of the task list, I'd like to:
- create a storybook entry for `Breadcrumbs` and iterate on it there
- Convert one of the smaller components first, like `RepositoryBranchesArea`

I'm still not sure about: how much existing CSS we should re-use for `Breadcrumbs` (e.g. `.repo-header-breadcrumb-nav-item`), or keep with the same class names but changed to match differences in Figma (if any), or use new classes from scratch based on the Figma styles.

**Comment by marekweb** [comment]
I converted `RepositoryBranchArea` with the assumption that it will receive `pushBreadcrumbs` as a prop. It's not receiving that prop yet so it won't currently work.

Next in order to provide `pushBreadcrumbs` it looks like `RepoContainer` needs to be converted to a function component as well, so that it can use the hook.

My guess about how to convert `RepoContainer` is to move all of its mount/unmount logic into `useEffect` but I'm not sure about the right way to deal with: all the subscriptions and `componentDidUpdate`, and also where/how it's appropriate to use `useObservable`. So I might delay on trying to convert it until I gain a better understanding.

**Comment by marekweb** [comment]
These are my additions to the storybook, as a rough draft to iterate on:

- Breadcrumbs: http://localhost:9001/?path=/story/web-breadcrumbs--three-levels-deep
- PageHeader: http://localhost:9001/?path=/story/web-pageheader--complex-header

I would need help to get unblocked on two fronts:

- What is the right Figma-to-CSS process here -- do we already have styles for the sub-components like the h1 or badge? For now in the PageHeader, I naively copied the header from the Insights Page so it's a starting point, but not yet styled according to the Figma designs.
- How do the observables/subscription patterns translate to function components? Namely I'm note sure what do we do with the subscription/subject objects which are instance properties (will `useObservable` be helpful here and how to use it?) and what to do with `componentDidUpdate`.

**Comment by lguychard** [comment]
I converted `<RepoContainer/>` to a function component to give an example of how that would work (converting `componentUpdates` patterns to simple `useEffect()` calls with props as dependencies, using `useObservable()` where applicable, etc.): https://github.com/sourcegraph/sourcegraph/pull/12552

**Comment by lguychard** [comment]
Regarding this:

> What is the right Figma-to-CSS process here -- do we already have styles for the sub-components like the h1 or badge? For now in the PageHeader, I naively copied the header from the Insights Page so it's a starting point, but not yet styled according to the Figma designs.

- Badge styles were updated by @felixfbecker in https://github.com/sourcegraph/sourcegraph/pull/12390
- For the header, let's try to compose as much as possible from existing utility classes (for margins, layout, etc.). I'm not sure about the `h1` -- if it doesn't match our existing styles for `.h1` / `<h1>`, does it match [web content styles](https://github.com/sourcegraph/sourcegraph/issues/12303)? Or is it different from both of those? If so, that might be something we'll want to review with design.

**Comment by felixfbecker** [comment]
> I converted a couple more pages, and noticed an issue with our useBreadcrumbs() hook: adding pushBreadcrumbs to the useEffect() dependencies array causes infinite render cycles.

Could this be prevented with `useMemo()` somehow to make `pushBreadcrumbs` a stable function reference?

**Comment by marekweb** [comment]
I made an attempt to make `Breadcrumbs` render in the repo header, to test the concept.

It renders but there's an issue with the breadcrumbs state, in which the old breadcrumbs don't get removed.

**Comment by lguychard** [comment]
fa57d0f fixes the "infinite loop if `pushBreadcrumb` is in dependencies" problem. It should also fix @marekweb's problem of breadcrumbs not getting removed.

**Comment by lguychard** [comment]
`RepoHeader` now contributes the repo name / repository dropdown through `pushBreadcrumb` instead of the contributions portal: 604cb3a 🎉 

Now working on converting RepositoryCompareArea.

**Comment by lguychard** [comment]
I got down a bit of a rabbithole converting `RepositoryCompareArea` to a function component (the hoverifier wouldn't cooperate). So -- for reference, here's how you use `pushBreadcrumb` in a class component 😄 : https://github.com/sourcegraph/sourcegraph/pull/12536/commits/c2c20374dc75412a8982fe76d0a862a6cf6ceba9

Also was still seeing duplicate branch selectors showing up -- fixed it with this cc @marekweb: https://github.com/sourcegraph/sourcegraph/pull/12536/commits/ac11ad28cd1e6c3b12108cc17855d32a947112b8

**Comment by eseliger** [comment]
@lguychard I got it to work in this component which is functional, maybe that helps? https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/web/src/enterprise/campaigns/detail/changesets/CampaignChangesets.tsx

**Comment by lguychard** [comment]
@eseliger my implementation is remarkably close to yours https://sourcegraph.com/github.com/sourcegraph/sourcegraph@4713f2b5acfa4c40cbdc4f4cfc4353a93ddb041a/-/blob/shared/src/hover/useHoverifier.ts?subtree=true

Position events are correctly detected on code views, but they are never resolved / `getHover()` is never called... Probably something dumb, I'll take another look at that branch soon.

@marekweb @felixfbecker the last Portal contribution to convert is the one in routes.tsx -- all other Repository* components now use `pushBreadcrumb()`

**Comment by lguychard** [comment]
Here is a fun one for the next of us who looks at this: toggling between tabs in settings messes up the order of the breadcrumbs:

Kapture 2020-07-31 at 15 05 42

So does the initial load of the Tags page.

**Comment by felixfbecker** [comment]
So I found a way to fix the order issue, but it's a bit of a weird solution because it depends on remembering elements forever: https://github.com/sourcegraph/sourcegraph/pull/12536/commits/2caefc8124617ec85748340308751c3ef9f2c76c

Thinking about a better solution, I think essentially the problem we're having here is that what we want to show is a _trace_ through the React component tree (where each node in the tree can register a node to be part of the trace), but sometimes nodes _in the middle_ can remove and add themselves back, messing up the order. I think it would be possible to solve this by modeling the API more like a tree (really a linked list, since every node only has one child) instead of just a stack: Having each component create a "breabcrumb node", which is passed to child components to register further breadcrumb sub-nodes. This would mean that when a component that is in the middle is changing its node, all the old breadcrumb descendants are discarded and descendant components are _forced_ to re-link to the new node object since it's passed as a prop and changed. Instead of doing this in `useEffect()`, it would be done in `useMemo()`:

```ts
const ParentComponent = ({ parentBreadcrumb }) => {
	const breadcrumb = useMemo(() => parentBreadcrumb.addBreadcrumb('key', <>Hello</>))
	return <ChildComponent parentBreadcrumb={breadcrumb} />
}
```

**Comment by lguychard** [comment]
Now using `setBreadcrumb()` in all `RepoContainer` children: https://github.com/sourcegraph/sourcegraph/pull/12536/commits/3d89207be5b3d0d51551b625321cf26bb41c1459

**Comment by tjkandala** [comment]
TODO to merge by EOW:
- [x] - fix unstable breadcrumb order
- [ ] - integration tests
- [ ] - wrap `Breadcrumbs` in a shared page header component

**Comment by felixfbecker** [comment]
I addded a lot of doc comments, @tjkandala could you check if I got them right?

I also aligned the `PageHeader` with the design, and moved the Breadcrumbs out of there, because not all pages that use breadcrumbs use that PageHeader and it would be difficult because the breadcrumbs should be full page width.

Made explore page and insights page use the PageHeader and match the design.

I think this is ready to go after adding the integration test @tjkandala mentioned above. @eseliger offered to adopt the new Breadcrumbs and PageHeader once it's in `main`, and we can tackle using it on the extension registry together with the registry improvements (it's a bit more entangled). There are more pages to adopt breadcrumbs (user settings etc) but it seems more valuable to get this into `main` as soon as possible.

**Comment by tjkandala** [comment]
The comments are great! I noticed that we forgot to pass `repoHeaderContributionsLifecycleProps` for the go to code host action, so changing that fixed the integration tests. I also added assertions for breadcrumb order.

**Comment by tjkandala** [comment]
Is the extra padding on the breadcrumbs intentional? If so, we should probably center the action buttons.
image

**Comment by felixfbecker** [comment]
Oh woops, that was not intentional.

**Comment by felixfbecker** [comment]
Fixed.

File: web/src/integration/repository.test.ts [hunk 1]
```diff
@@ -368,8 +368,23 @@ describe('Repository', () => {
             await driver.page.waitForSelector('.test-repo-blob')
             await driver.assertWindowLocation(`${repositorySourcegraphUrl}/-/blob/${clickedFileName}`)
 
-            // Assert that the file is loaded
-            await assertSelectorHasText('.breadcrumb .part-last', clickedFileName)
+            // Assert breadcrumb order
+            await driver.page.waitForSelector('[aria-label="Breadcrumbs"]')
```
Comment (felixfbecker) [inline web/src/integration/repository.test.ts:372]: **issue**: we should use stable `test-` identifiers so tests don't break when e.g. the ARIA label is changed. See https://docs.sourcegraph.com/dev/testing#finding-elements-with-css-selectors

File: web/src/integration/repository.test.ts [hunk 2]
```diff
@@ -368,8 +368,23 @@ describe('Repository', () => {
...
+                return textContents
+            })
+            const expectedBreadCrumbsText = ['Home', 'Repositories', shortRepositoryName, '@master', clickedFileName]
+
+            for (const [index, textContent] of breadcrumbsText.entries()) {
+                assert.strictEqual(textContent, expectedBreadCrumbsText[index])
+            }
+
+            // // Assert that the file is loaded
+            // await assertSelectorHasText('.breadcrumb .part-last', clickedFileName)
```
Comment (felixfbecker) [inline web/src/integration/repository.test.ts:387]: **issue**: should this be removed?

File: web/src/integration/repository.test.ts [hunk 3]
```diff
@@ -368,8 +368,23 @@ describe('Repository', () => {
...
+                document
+                    .querySelector('[aria-label="Breadcrumbs"]')
+                    ?.childNodes.forEach(node => node.textContent && textContents.push(node.textContent))
+                return textContents
+            })
+            const expectedBreadCrumbsText = ['Home', 'Repositories', shortRepositoryName, '@master', clickedFileName]
+
+            for (const [index, textContent] of breadcrumbsText.entries()) {
+                assert.strictEqual(textContent, expectedBreadCrumbsText[index])
+            }
```
Comment (felixfbecker) [inline web/src/integration/repository.test.ts:384]: **suggestion**: `assert.deepStrictEqual()` would give more debuggable output if this fails. See https://about.sourcegraph.com/handbook/engineering/web/testing#use-specific-assertion-functions

File: web/src/Layout.tsx [hunk 4]
```diff
@@ -144,12 +145,19 @@ export const Layout: React.FunctionComponent<LayoutProps> = props => {
...
 
+    const breadcrumbProps = useBreadcrumbs()
+
     useScrollToLocationHash(props.location)
     // Remove trailing slash (which is never valid in any of our URLs).
     if (props.location.pathname !== '/' && props.location.pathname.endsWith('/')) {
         return <Redirect to={{ ...props.location, pathname: props.location.pathname.slice(0, -1) }} />
     }
 
+    const context = {
```
[the remark this answers is not in the document]
Comment (felixfbecker) [reply web/src/Layout.tsx:157]: I agree with `context` being overused (and often misused). That said, we use `context` very consistently for the _context_ passed by routing components into each subroute, and I think it's a the most fitting name for that.
An example where I think `context` is not fitting is when it's simply used as an options/arguments object bag for a pure function (the passed data doesn't necessarily have to be the "context" the function is called in).

**Pull request merged**